src/check/parse/tokenize/src/main.zig

```diff
@@ -0,0 +1,1399 @@
+pub const Cursor = struct {
+    buf: []const u8,
+    pos: usize,
+    messages: []Message,
+    message_count: usize,
+    gc: *GenCatData,
+
+    /// Initialize a Cursor with the given input buffer and a pre-allocated messages slice.
+    pub fn init(buf: []const u8, messages: []Message, gc: *GenCatData) Cursor {
+        return Cursor{
+            .buf = buf,
+            .pos = 0,
+            .messages = messages,
+            .message_count = 0,
+            .gc = gc,
+        };
+    }
+
+    fn pushMessageHere(self: *Cursor, kind: MessageKind) void {
+        if (self.message_count < self.messages.len) {
+            self.messages[self.message_count] = Message{
+                .kind = kind,
+                .begin = @intCast(self.pos),
+                .end = @intCast(self.pos),
+            };
+            self.message_count += 1;
+        }
+    }
+
+    fn pushMessage(self: *Cursor, kind: MessageKind, begin: u32, end: u32) void {
+        if (self.message_count < self.messages.len) {
+            self.messages[self.message_count] = Message{
+                .kind = kind,
+                .begin = begin,
+                .end = end,
+            };
+            self.message_count += 1;
+        }
+    }
+
+    /// Returns the current byte, or null if at the end.
+    pub fn peek(self: *Cursor) ?u8 {
+        if (self.pos < self.buf.len) {
+            return self.buf[self.pos];
+        }
+        return null;
+    }
+
+    /// Returns the byte at the given lookahead offset.
+    pub fn peekAt(self: *Cursor, lookahead: usize) ?u8 {
+        if (self.pos + lookahead < self.buf.len) {
+            return self.buf[self.pos + lookahead];
+        }
+        return null;
+    }
+
+    pub fn isPeekedCharInRange(self: *Cursor, lookahead: usize, start: u8, end: u8) bool {
+        if (self.pos + lookahead < self.buf.len) {
+            const ch = self.buf[self.pos + lookahead];
+            return ch >= start and ch <= end;
+        }
+        return false;
+    }
+
+    /// Requires that the next byte is `ch`, otherwise pushes a message.
+    pub fn require(self: *Cursor, ch: u8, kind: MessageKind) void {
+        if (self.peek() == ch) {
+            self.pos += 1;
+        } else {
+            self.pushMessageHere(kind);
+        }
+    }
+
+    /// Chomps “trivia” (whitespace, comments, etc.) and returns an optional indent.
+    pub fn chompTrivia(self: *Cursor) ?Indent {
+        var sawNewline = false;
+        var indent = Indent.init();
+
+        while (self.pos < self.buf.len) {
+            const b = self.buf[self.pos];
+            if (b == ' ') {
+                self.pos += 1;
+                if (sawNewline) indent.num_spaces += 1;
+            } else if (b == '\t') {
+                self.pos += 1;
+                if (sawNewline) indent.num_tabs += 1;
+            } else if (b == '\n') {
+                self.pos += 1;
+                sawNewline = true;
+                indent = Indent.init();
```

Comment:
Why reinitialize the ident? It is still just `Ident.init()` like at the top of the function.

Reply:
Yeah, I think this is simpler with Indent, Dedent, and Newline tokens.

Reply:
See this thread: https://github.com/roc-lang/roc/pull/7569#discussion_r1938606858